mediapipe/tasks/ios/genai/inference/sources/LlmInference+Session.swift

```diff
@@ -0,0 +1,289 @@
+// Copyright 2024 The MediaPipe Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import MediaPipeTasksGenAIC
+
+extension LlmInference {
+
+  /// An `LlmInference` Session that can be used to execute queries using the LLM that was used to
+  /// initialize the `LlmInference`task.
+  /// You can create multiple query sessions using the same `LlmInference` which can be active at
+  /// the same time.
+  /// You can also clone an existing session and continue querying the LLM from where you left off.
+  ///
+  /// Note: Inherits from `NSObject` for Objective C interoperability.
+  @objc(MPPLLMInferenceSession) public final class Session: NSObject {
+    private static let responseGenerationInProgressQueueName =
+      "com.google.mediapipe.genai.isResponseGenerationInProgressQueue"
+
+    // Session runner that manages the creation, deletion and execution of the underlying C session.
+    private let llmSessionRunner: LlmSessionRunner
+
+    // Queue that restricts access to the response generation functions simultaneously.
+    private let responseGenerationInProgressQueue = DispatchQueue(
+      label: Session.responseGenerationInProgressQueueName,
+      attributes: .concurrent)
+
+    /// Tracks whether a response generation is in progress.
+    /// Readers writers lock to prevent race condition as this variable can be accessed from multiple
+    /// threads.
+    private var responseGenerationInProgressInternal = false
+    private var responseGenerationInProgress: Bool {
+      get {
+        responseGenerationInProgressQueue.sync {
+          return self.responseGenerationInProgressInternal
+        }
+      }
+      set {
+        responseGenerationInProgressQueue.async(flags: .barrier) {
+          self.responseGenerationInProgressInternal = newValue
+        }
+      }
+    }
+
+    /// Creates a new instance of `LlmInference` with the given options. 
+    /// Ensure that the instance of `LlmInference` passed to this function must be alive through 
+    /// the lifetime of the current session. Using a session beyond the life time of the 
+    /// `LlmInference` used to create it will lead to undefined behaviour.
```

Comment:
@yishuangP I decided to not store the llmInference in the session since it is not used in any API other than creation of the Session.  But to avoid the llmInference getting deallocated before all the sessions created from it are destroyed, we can store it in the session as an instance variable. Won't lead to a retain cycle since `LlmInference` does not have any reference to `Session`. Just that the `llmInference` may remain in memory even if it goes out of scope in the user program until all sessions are destroyed. LMK your thoughts. We can do this in a follow up PR if needed. Not critical at the moment.